File: backend/app/schemas/publisher.py
from pydantic import BaseModel, ConfigDict, Field
from .product import PriorityMapSchema


class SubmissionCreate(BaseModel):
    # Basic info (all optional for draft saving)
    name: str | None = None
    type: str | None = None
    tagline: str | None = None
    description: str | None = None
    url: str | None = None
    logo: str | None = None
    tags: list[str] = []
    audience: list[str] = []
    problems_solved: list[str] = []
    audience_priorities: PriorityMapSchema = PriorityMapSchema()
    problems_priorities: PriorityMapSchema = PriorityMapSchema()
    pricing_model: str | None = None
    pricing_amount: str | None = None
    pricing_details: str | None = None
    protocol_answers: dict = {}


class SubmissionUpdate(BaseModel):
    # Same as Create but explicit partial
    name: str | None = None
    type: str | None = None
    tagline: str | None = None
    description: str | None = None
    url: str | None = None
    logo: str | None = None
    tags: list[str] | None = None
    audience: list[str] | None = None
    problems_solved: list[str] | None = None
    audience_priorities: PriorityMapSchema | None = None
    problems_priorities: PriorityMapSchema | None = None
    pricing_model: str | None = None
    pricing_amount: str | None = None
    pricing_details: str | None = None
    protocol_answers: dict | None = None


class SubmissionResponse(BaseModel):
    id: str
    publisherId: str
    productId: str | None
    status: str
    name: str | None
    type: str | None
    tagline: str | None
    description: str | None
    url: str | None
    logo: str | None
    tags: list[str]
    audience: list[str]
    problemsSolved: list[str]
    audiencePriorities: dict | None = None
    problemsPriorities: dict | None = None
    pricingModel: str | None
    pricingAmount: str | None
    pricingDetails: str | None
    protocolAnswers: dict
    adminNotes: str | None
    reviewedAt: str | None
    createdAt: str
    updatedAt: str

    model_config = ConfigDict(from_attributes=True)


class AdminReviewAction(BaseModel):
    admin_notes: str | None = None
    # Link to an already-created product instead of creating from submission data
    product_id: str | None = None
    # For approve: also set scores on the product
    score_security: int | None = Field(default=None, ge=0, le=5)
    score_efficacy: int | None = Field(default=None, ge=0, le=5)
    score_accessibility: int | None = Field(default=None, ge=0, le=5)
    score_ux: int | None = Field(default=None, ge=0, le=5)
    score_support: int | None = Field(default=None, ge=0, le=5)
    justification_security: str | None = None
    justification_efficacy: str | None = None
    justification_accessibility: str | None = None
    justification_ux: str | None = None
    justification_support: str | None = None


class AdminCreateAndPublishSchema(BaseModel):
    # Basic product info (same as SubmissionCreate)
    id: str | None = None
    name: str
    type: str
    tagline: str
    description: str
    url: str
    logo: str | None = None
    tags: list[str] = []
    audience: list[str] = []
    problems_solved: list[str] = []
    audience_priorities: PriorityMapSchema = PriorityMapSchema()
    problems_priorities: PriorityMapSchema = PriorityMapSchema()
    is_mentaltech_member: bool = False
    preference_match: list[str] = []
    pricing_model: str | None = None
    pricing_amount: str | None = None
    pricing_details: str | None = None
    protocol_answers: dict = {}
    last_updated: str | None = None
    # Scores (admin sets directly)
    score_security: int | None = Field(default=None, ge=0, le=5)
    score_efficacy: int | None = Field(default=None, ge=0, le=5)
    score_accessibility: int | None = Field(default=None, ge=0, le=5)
    score_ux: int | None = Field(default=None, ge=0, le=5)
    score_support: int | None = Field(default=None, ge=0, le=5)
    justification_security: str | None = None
    justification_efficacy: str | None = None
    justification_accessibility: str | None = None
    justification_ux: str | None = None
    justification_support: str | None = None
    scoring_criteria: dict | None = None

File: frontend/src/api/publisher.ts
import { apiFetch } from "./client";
import type { Product } from "../types";

// Admin functions
export async function listSubmissions(
  status?: string,
): Promise<Record<string, unknown>[]> {
  const query = status ? `?status=${status}` : "";
  return apiFetch<Record<string, unknown>[]>(
    `/admin/submissions${query}`,
    {},
    true,
  );
}

export async function getAdminSubmission(
  id: string,
): Promise<Record<string, unknown>> {
  return apiFetch<Record<string, unknown>>(
    `/admin/submissions/${id}`,
    {},
    true,
  );
}

export async function approveSubmission(
  id: string,
  reviewData: Record<string, unknown>,
): Promise<Record<string, unknown>> {
  return apiFetch<Record<string, unknown>>(
    `/admin/submissions/${id}/approve`,
    {
      method: "POST",
      body: JSON.stringify(reviewData),
    },
    true,
  );
}

export async function rejectSubmission(
  id: string,
  adminNotes: string,
): Promise<Record<string, unknown>> {
  return apiFetch<Record<string, unknown>>(
    `/admin/submissions/${id}/reject`,
    {
      method: "POST",
      body: JSON.stringify({ admin_notes: adminNotes }),
    },
    true,
  );
}

export async function requestChanges(
  id: string,
  adminNotes: string,
): Promise<Record<string, unknown>> {
  return apiFetch<Record<string, unknown>>(
    `/admin/submissions/${id}/request-changes`,
    {
      method: "POST",
      body: JSON.stringify({ admin_notes: adminNotes }),
    },
    true,
  );
}

export async function createAndPublishAdmin(data: {
  id?: string;
  name: string;
  type: string;
  tagline: string;
  description: string;
  url: string;
  logo?: string;
  tags: string[];
  audience: string[];
  problems_solved: string[];
  preference_match: string[];
  is_mentaltech_member: boolean;
  pricing_model?: string;
  pricing_amount?: string;
  pricing_details?: string;
  protocol_answers: Record<string, unknown>;
  score_security?: number;
  score_efficacy?: number;
  score_accessibility?: number;
  score_ux?: number;
  score_support?: number;
  justification_security?: string;
  justification_efficacy?: string;
  justification_accessibility?: string;
  justification_ux?: string;
  justification_support?: string;
  scoring_criteria?: Record<string, unknown>;
  audience_priorities?: Record<string, string[]>;
  problems_priorities?: Record<string, string[]>;
}): Promise<Product> {
  return apiFetch<Product>(
    "/admin/submissions/create-and-publish",
    {
      method: "POST",
      body: JSON.stringify(data),
    },
    true,
  );
}

export async function uploadLogoAdmin(file: File): Promise<string> {
  const formData = new FormData();
  formData.append("file", file);
  const result = await apiFetch<{ path: string }>(
    "/admin/upload-logo",
    {
      method: "POST",
      body: formData,
      headers: {},
    },
    true,
  );
  return result.path;
}
